File: BinarySearchTree.py
class TreeRoot():
    def __init__(self, x):
        self.data = x
        self.left = None
        self.right = None

class BinarySearchTree():
    def Search(self, root, x):
        if root == None:
            return False
        if x < root.data:
            return self.Ssearch(root.left, x, root)
        if x > root.data:
            return self.Search(root.right, x, root)
        return True

    def Insert(self, root, x):
        if root == None:
            root = TreeRoot(x)
            return root
        if x < root.data:
            root.left = self.Insert(root.left, x)
        else:
            root.right = self.Insert(root.right, x)
        return root

    def Delete(self, root, x):
        if root == None:
            return None
        if x < root.data:
            root.left = self.Delete(root.left, x)
        if x > root.data:
            root.right = self.Delete(root.right, x)
            
        if root.left == None:
            root = root.right
            return root

        if root.right == None:
            root = root.left
            return root

        tmp = root.left
        while tmp.right:
            tmp = tmp.right
        root.data = tmp.data
        root.left = self.Delete(root.left, root.data)
        return root

    def ListRoot(self, root):
        if root:
            return self.ListRoot(root.left) + [root.data] + self.ListRoot(root.right)
        return []

if __name__ == '__main__':
    Tree = BinarySearchTree()
    root = None
    list = [3, 4, 1, 2]
    for i in range(len(list)):
        root = Tree.Insert(root, list[i])
    root = Tree.Delete(root, 2)
    print(Tree.ListRoot(root))

File: BinarySearchTree.cpp
#include<iostream>

struct TreeRoot {
    int data;
    TreeRoot *left = NULL, *right = NULL;
};

TreeRoot *Insert(TreeRoot *root, int x) {
    if (root == NULL) {
        root = new TreeRoot;
        root->data = x;
        return root;
    }
    if (x < root->data)
        root->left = Insert(root->left, x);
    else
        root->right = Insert(root->right, x);
    return root;
}

bool Search(TreeRoot *root, int x) {
    if (root == NULL)
        return false;
    if (x < root->data)
        return Search(root->left, x);
    if (x > root->data)
        return Search(root->right, x);
    return true;
}

TreeRoot *Delete(TreeRoot *root, int x) {
    if (root == NULL)
        return NULL;
    if (x < root->data)
        root->left = Delete(root->left, x);
	if (x > root->data)
        root->right = Delete(root->right, x);
    
    if (root->left == NULL) {
        TreeRoot *tmp = root->right;
        delete root;
        return tmp;
    }

    if (root->right == NULL) {
        TreeRoot *tmp = root->left;
        delete root;
        return tmp;
    }

    TreeRoot *tmp;
    for (tmp = root->left; tmp->right != NULL; tmp = tmp->right);

    root->data = tmp->data;
    root->left = Delete(root->left, root->data);
    return root;
}

void PrintTree(TreeRoot *root)
{
    if(root != NULL)
    {
        PrintTree(root->left);
        std::cout << ' ' << root->data << ' ';
        PrintTree(root->right);
    }
    
}

int main(void)
{
    TreeRoot *root = NULL;
    root = Insert(root, 3);
    root = Insert(root, 4);
    root = Insert(root, 1);
    root = Insert(root, 2);
    PrintTree(root);
    return 0;
}